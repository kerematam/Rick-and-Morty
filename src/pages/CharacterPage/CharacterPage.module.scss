.back_button {
  width: 100%;
}

.detail_card {
  font-family: Roboto;
  width: 300px;
  margin-bottom: 20px;
  .detail_item {
    margin: 4px;
    display: flex;
    justify-content: space-between;
  }
  .detail_item_key {
    font-weight: 500;
    margin-left: 5px;
  }
  .detail_item_value {
    margin-left: 5px;
    margin-right: 5px;
  }
}

.detail_card_container {
  margin-right: 30px;
  margin-bottom: 20px;
}

.table_container {
  padding-top: 30px;
  max-width: 400px;
}

.page_container {
  flex-wrap: wrap;
  display: flex;
}
