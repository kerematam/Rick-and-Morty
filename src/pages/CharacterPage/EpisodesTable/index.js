import React from 'react'
import PropTypes from 'prop-types'
import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper,
  LinearProgress,
} from '@material-ui/core'

import styles from './EpisodesTable.module.scss'

const EpisodesTable = React.memo(({ episodes, loading }) => {
  return (
    <TableContainer component={Paper}>
      {loading && <LinearProgress></LinearProgress>}
      <Table className={styles.table} aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell>Episode </TableCell>
            <TableCell>Name</TableCell>
            <TableCell>Air Date</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {episodes.map(({ id, episode, name, air_date, url }) => (
            <TableRow hover key={id}>
              <TableCell component="th" scope="row">
                {episode}
              </TableCell>
              <TableCell>{name}</TableCell>
              <TableCell>{air_date}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  )
})

EpisodesTable.propTypes = {
  episodes: PropTypes.array,
  loading: PropTypes.bool,
}

export default EpisodesTable
