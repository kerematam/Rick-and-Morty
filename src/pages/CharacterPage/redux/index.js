import { useInjectReducer, useInjectSaga } from 'redux-injectors'
import saga from './sagas'
import { createSlice } from '@reduxjs/toolkit'

export const NAMESPACE = 'CharacterDetailPage'

export const initialState = {
  loading: false,
  character: {},
  episodes: [],
}

const slice = createSlice({
  name: NAMESPACE,
  initialState,
  reducers: {
    loadCharacter: state => {
      state.loading = true
    },
    loadCharacterSucceed: (state, action) => {
      state.character = action.payload.character
      state.loading = false
      state.error = false
    },
    setActiveCharacter: (state, action) => {
      state.character = action.payload.character
      state.loading = false
      state.error = false
      state.episodes = []
    },
    loadCharacterFailed: (state, action) => {
      state.loading = false
      state.error = true
    },
    loadEpisodes: state => {
      state.episodesLoading = true
    },
    loadEpisodesSucceed: (state, action) => {
      state.episodesLoading = false
      state.episodes = action.payload.data
    },
    loadEpisodesFailed: state => {
      state.episodesLoading = false
    },
  },
})

export const { reducer, actions } = slice

// Injectors
export const useCharacterDetailReducerInjector = () =>
  useInjectReducer({ key: NAMESPACE, reducer })
export const useCharacterDetailSagaInjector = () =>
  useInjectSaga({ key: NAMESPACE, saga })
export const useCharacterDetailReduxInjector = () => {
  useCharacterDetailReducerInjector()
  useCharacterDetailSagaInjector()
}

// Selectors
export const characterDetailSelector = state => state[NAMESPACE]
