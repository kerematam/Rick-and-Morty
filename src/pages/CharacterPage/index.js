import React, { useEffect, useCallback } from 'react'
import { Helmet } from 'react-helmet-async'
import EpisodesTable from './EpisodesTable'
import { useSelector, useDispatch } from 'react-redux'
import { useRouteMatch } from 'react-router-dom'
import {
  Paper,
  Divider,
  Typography,
  Button,
  LinearProgress,
} from '@material-ui/core/'
import ArrowBackIcon from '@material-ui/icons/ArrowBack'
import {
  actions,
  useCharacterDetailReduxInjector,
  characterDetailSelector,
} from './redux'
import { useCharacterReduxInjector } from 'pages/CharacterListPage/redux'
import history from 'utils/history'

import styles from './CharacterPage.module.scss'

/**
 *  Returns character and episodes
 *  if redirected from from list page, it loads from store
 *  if directly accessed via URL; load action is triggered
 */
const useSelectCharacter = () => {
  useCharacterReduxInjector()
  useCharacterDetailReduxInjector()
  const { character, loading, episodes, episodesLoading } = useSelector(
    characterDetailSelector
  )
  const dispatch = useDispatch()

  const {
    params: { id },
  } = useRouteMatch()

  useEffect(() => {
    const { loadCharacter, loadEpisodes } = actions
    if (!character.id) {
      dispatch(loadCharacter({ id }))
    } else {
      dispatch(loadEpisodes({ episodeUrls: character.episode }))
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps,
  }, [])

  return [character, loading, episodes, episodesLoading]
}

const useSelectPreviousRoute = () => {
  const routes = useSelector(state => state.routerLocations)
  return routes.length > 1 && routes.slice(-2)[0]
}

const CharacterPage = () => {
  const [
    character,
    characterLoading,
    episodes,
    episodesLoading,
  ] = useSelectCharacter()

  const previousRoute = useSelectPreviousRoute()

  const goBack = useCallback(() => {
    const pathname = previousRoute?.location?.pathname
    const isPreviousHomePage =
      pathname === '/characters/' || pathname === '/characters'

    if (isPreviousHomePage) {
      history.goBack()
    } else {
      history.push('/characters')
    }
  }, [previousRoute])

  return (
    <div>
      <Helmet>
        <title>{character?.name}</title>
      </Helmet>
      <div className={styles.page_container}>
        <div className={styles.detail_card_container}>
          <Paper className={styles.detail_card}>
            <Button
              onClick={goBack}
              className={styles.back_button}
              startIcon={<ArrowBackIcon />}
            >
              Characters Page
            </Button>
          </Paper>
          <Paper className={styles.detail_card}>
            {characterLoading && <LinearProgress></LinearProgress>}
            <Typography className={styles.detail_item}>
              <span className={styles.detail_item_key}>Name</span>
              <span className={styles.detail_item_value}>
                {character?.name}
              </span>
            </Typography>
            <Divider />
            <Typography className={styles.detail_item}>
              <span className={styles.detail_item_key}>Status</span>
              <span className={styles.detail_item_value}>
                {character?.status}
              </span>
            </Typography>
            <Divider />
            <Typography className={styles.detail_item}>
              <span className={styles.detail_item_key}>Origin</span>
              <span className={styles.detail_item_value}>
                {character?.origin?.name}
              </span>
            </Typography>
            <Divider />
          </Paper>
          <img alt="Character" src={character?.image} />
        </div>
        <div>
          <EpisodesTable
            className={styles.table}
            episodes={episodes}
            loading={episodesLoading || characterLoading}
          />
        </div>
      </div>
    </div>
  )
}

export default CharacterPage
