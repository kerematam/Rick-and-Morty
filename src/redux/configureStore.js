import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import createReducer from './rootReducers'
import { routerMiddleware } from 'connected-react-router'
import { createInjectorsEnhancer, forceReducerReload } from 'redux-injectors'

function configureStore(initialState = {}, history) {
  let composeEnhancers = compose
  const reduxSagaMonitorOptions = {}

  if (process.env.NODE_ENV !== 'production' && typeof window === 'object') {
    if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__)
      composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
  }

  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions)
  const { run: runSaga } = sagaMiddleware
  const middlewares = [sagaMiddleware, routerMiddleware(history)]

  const enhancers = [
    applyMiddleware(...middlewares),
    createInjectorsEnhancer({
      createReducer,
      runSaga,
    }),
  ]

  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(...enhancers)
  )

  if (module.hot) {
    module.hot.accept('./rootReducers', () => {
      forceReducerReload(store)
    })
  }

  return store
}

export default configureStore
