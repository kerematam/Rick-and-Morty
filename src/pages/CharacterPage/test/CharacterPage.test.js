import React, { useCallback } from 'react'
import renderer from 'react-test-renderer'
import CharacterPage from '../index'
import { Provider } from 'react-redux'
import configureStore from 'redux/configureStore'
import { HelmetProvider } from 'react-helmet-async'
import { createMemoryHistory } from 'history'
import { Router } from 'react-router'

const setup = () => {
  const store = configureStore()
  const history = createMemoryHistory()

  const wrapper = renderer
    .create(
      <Provider store={store}>
        <Router history={history}>
          <HelmetProvider>
            <CharacterPage />
          </HelmetProvider>
        </Router>
      </Provider>
    )
    .toJSON()

  return {
    wrapper,
  }
}

describe('CharacterPage', () => {
  test('should render correctly', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })
})
