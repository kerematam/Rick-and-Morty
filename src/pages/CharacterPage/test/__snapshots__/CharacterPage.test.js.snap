// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterPage should render correctly 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-flex-start"
    flexShrink="0"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
    >
      <div
        className="MuiGrid-root page_left_column MuiGrid-container"
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <div
            className="MuiPaper-root detail_card MuiPaper-elevation1 MuiPaper-rounded"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text back_button"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                    />
                  </svg>
                </span>
                Characters Page
              </span>
            </button>
          </div>
          <div
            className="MuiPaper-root detail_card MuiPaper-elevation1 MuiPaper-rounded"
          >
            <p
              className="MuiTypography-root detail_item MuiTypography-body1"
            >
              <span
                className="detail_item_key"
              >
                Name
              </span>
              <span
                className="detail_item_value"
              />
            </p>
            <hr
              className="MuiDivider-root"
            />
            <p
              className="MuiTypography-root detail_item MuiTypography-body1"
            >
              <span
                className="detail_item_key"
              >
                Status
              </span>
              <span
                className="detail_item_value"
              />
            </p>
            <hr
              className="MuiDivider-root"
            />
            <p
              className="MuiTypography-root detail_item MuiTypography-body1"
            >
              <span
                className="detail_item_key"
              >
                Origin
              </span>
              <span
                className="detail_item_value"
              />
            </p>
            <hr
              className="MuiDivider-root"
            />
          </div>
        </div>
        <div
          className="MuiGrid-root background_blue"
        >
          <div
            className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <img
              alt="Character"
              className="img_detail"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root background_red MuiGrid-item MuiGrid-grid-xs-9"
    >
      <div
        className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <table
          aria-label="simple table"
          className="MuiTable-root table"
        >
          <thead
            className="MuiTableHead-root"
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head"
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Episode 
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Name
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Air Date
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root"
          />
        </table>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    />
  </div>
</div>
`;
